AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageTest.java

```diff
@@ -0,0 +1,111 @@
+    setupSDK(mContext, "/responses/simple_start_response.json");
+  }
+
+  @Test
+  public void testMessageCreate() {
+    Date delivery = new Date(100);
+    Date expiration = new Date(200);
+    HashMap<String, Object> map = new HashMap<>();
+    map.put(Constants.Keys.MESSAGE_DATA, new HashMap<String, Object>());
+    map.put(Constants.Keys.DELIVERY_TIMESTAMP, delivery.getTime());
+    map.put(Constants.Keys.EXPIRATION_TIMESTAMP, expiration.getTime());
+    map.put(Constants.Keys.IS_READ, true);
+
+    LeanplumInboxMessage message = LeanplumInboxMessage.createFromJsonMap("message##Id", map);
+    assertEquals("message##Id", message.getMessageId());
+    assertEquals(delivery, message.getDeliveryTimestamp());
+    assertEquals(expiration, message.getExpirationTimestamp());
+    assertTrue(message.isRead());
+    assertNull(message.getData());
+
+    assertNull(message.getImageFilePath());
+    assertNull(message.getImageUrl());
+  }
+
+  @Test
+  public void testReadAndUpdateMessageCount() {
+    Date delivery = new Date(100);
+    Date expiration = new Date(200);
+    HashMap<String, Object> map = new HashMap<>();
+    map.put(Constants.Keys.MESSAGE_DATA, new HashMap<String, Object>());
+    map.put(Constants.Keys.DELIVERY_TIMESTAMP, delivery.getTime());
+    map.put(Constants.Keys.EXPIRATION_TIMESTAMP, expiration.getTime());
+    map.put(Constants.Keys.IS_READ, false);
+    LeanplumInboxMessage message = LeanplumInboxMessage
+        .createFromJsonMap("messageId##00", map);
+    int intialUnreadCount = leanplumInbox.unreadCount();
+
+    assertEquals(false, message.isRead());
+
+    message.read();
+
+    assertEquals(true, message.isRead());
+    assertEquals(intialUnreadCount-1, leanplumInbox.unreadCount());
+  }
+
+  @Test
+  public void testImageURL() {
+    leanplumInbox.disableImagePrefetching();
+    ResponseHelper.seedResponse("/responses/newsfeed_response.json");
+    leanplumInbox.downloadMessages();
+    List<LeanplumInboxMessage> messagesList = leanplumInbox.allMessages();
+    LeanplumInboxMessage imageMessage = messagesList.get(0);
+
+    String actualUrl = imageMessage.getImageUrl().toString();
+
+    assertEquals("http://bit.ly/2GzJxxx",
+        actualUrl);
+  }
+
+  @Test
+  public void testImageFilepathIsReturnedIfPrefetchingEnabled() {
+    ResponseHelper.seedResponse("/responses/newsfeed_response.json");
+    leanplumInbox.downloadMessages();
+    LeanplumInboxMessage imageMessage = leanplumInbox.allMessages().get(0);
+
+    String imageFilePath = imageMessage.getImageFilePath();
+
+    assertNotNull(imageFilePath);
+  }
+
+  @Test
+  public void testOpenAction() {
+    ResponseHelper.seedResponse("/responses/newsfeed_response.json");
+    leanplumInbox.downloadMessages();
+    LeanplumInboxMessage imageMessage = leanplumInbox.allMessages().get(1);
+    imageMessage.read();
+
+    HashMap actionName = imageMessage.getContext().objectNamed("Open action");
+
+    assertEquals(true, actionName.containsValue("Alert"));
```

Comment:
actionName.containsValue must be equals "Alert"

Reply:
actionName is a hashmap which looks like {"__name__": "Alert"}, so when we call assertEquals(true, actionName.containsValue("Alert")) it is the equivalent of doing assertEquals(actionName.get("__name__"), "Alert");

i personally dont have a preference, I can switch it to the second way for better readability.